### PR TITLE
adding implementation for replica entity id

### PR DESCRIPTION
**Pull request by abhilashshetty04**
This PR allows control plane to send entity_id (VolumeId) in CreateReplicaRequest. 
For the older replicas. If entity_id is None in Replica object, it sets it by calling a newly added grpc call.

### CODE REVIEW

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

**Comment by abhilashshetty04** [comment]
bors try

File: control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs [hunk 1]
```diff
@@ -90,6 +91,15 @@ impl TryIoEngineToAgent for v1::replica::Replica {
                 uuid: self.uuid.to_owned(),
                 kind: ResourceKind::Replica,
             })?,
+            entity_id: match &self.entity_id {
+                None => None,
+                Some(owner_id) => Some(VolumeId::try_from(owner_id.clone()).map_err(|_| {
+                    SvcError::InvalidUuid {
+                        uuid: "".to_string(),
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs]:
```suggestion
                        uuid: owner_id,
```

File: control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs [hunk 2]
```diff
@@ -90,6 +91,15 @@ impl TryIoEngineToAgent for v1::replica::Replica {
                 uuid: self.uuid.to_owned(),
                 kind: ResourceKind::Replica,
             })?,
+            entity_id: match &self.entity_id {
+                None => None,
+                Some(owner_id) => Some(VolumeId::try_from(owner_id.clone()).map_err(|_| {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs]: Does the dataplane enforce the owner it to be uuid v4?

File: control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs [hunk 3]
```diff
@@ -259,7 +269,7 @@ impl AgentToIoEngine for transport::CreateReplica {
         Self::IoEngineMessage {
             name: ReplicaName::from_opt_uuid(self.name.as_ref(), &self.uuid).into(),
             uuid: self.uuid.clone().into(),
-            entity_id: None,
+            entity_id: self.entity_id.clone().map(|id| id.into()),
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs]:
```suggestion
            entity_id: self.entity_id.into_opt(),
```

File: control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs [hunk 4]
```diff
@@ -427,7 +427,7 @@ pub(super) async fn missing_nexus_recreate(
         for item in children.candidates() {
             // just in case the replica gets somehow shared/unshared?
             match nexus_guard
-                .make_me_replica_accessible(context.registry(), item.state())
+                .make_me_replica_accessible(context.registry(), item.state(), nexus.owner.clone())
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/reconciler/nexus/mod.rs]: Do we need to pass the nexus owner? The replica already has a volume owner, why don't we just use that?

File: control-plane/agents/src/bin/core/controller/resources/operations.rs [hunk 5]
```diff
@@ -230,6 +232,13 @@ pub(crate) trait ResourceOwnerUpdate {
         // pre-update the actual spec anyway since this is a removal,
         update_on_commit: bool,
     ) -> Result<(), SvcError>;
+
+    /// Set entity id on the replica.
+    async fn set_entity_id(
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/resources/operations.rs]: hmm not needed, replica spec already has owners we can use

File: control-plane/agents/src/bin/core/nexus/operations_helper.rs [hunk 6]
```diff
@@ -132,7 +135,19 @@ impl OperationGuardArc<NexusSpec> {
         registry: &Registry,
         replica_state: &Replica,
         nexus_node: &NodeId,
+        volume_id: Option<VolumeId>,
     ) -> Result<ChildUri, SvcError> {
+        if replica_state.entity_id.is_none() {
+            if let Some(vol) = volume_id {
+                let mut replica = registry.specs().replica(&replica_state.uuid).await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/nexus/operations_helper.rs]:
```suggestion
            let mut replica = registry.specs().replica(&replica_state.uuid).await?;
            if let Some(vol) = replica.owners.volume.as_ref() {
```

File: control-plane/agents/src/bin/core/node/wrapper.rs [hunk 7]
```diff
@@ -1393,6 +1393,18 @@ impl ReplicaApi for Arc<tokio::sync::RwLock<NodeWrapper>> {
...
+
+    /// Set entity id for a given replica via grpc.
+    async fn set_replica_entity_id(
+        &self,
+        request: &SetReplicaEntityId,
+    ) -> Result<Replica, SvcError> {
+        let dataplane = self.grpc_client_locked(request.id()).await?;
+        let replica = dataplane.set_replica_entity_id(request).await?;
+        let mut ctx = dataplane.reconnect(GETS_TIMEOUT).await?;
+        self.update_replica_states(ctx.deref_mut()).await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/wrapper.rs]: Doesn't seem right.. should only update this replica

File: control-plane/stor-port/src/types/v0/transport/replica.rs [hunk 8]
```diff
@@ -556,6 +558,40 @@ impl From<ReplicaOwners> for models::ReplicaSpecOwners {
     }
 }
 
+/// Request struct to set replica entity id.
+#[derive(Serialize, Deserialize, Default, Debug, Clone, PartialEq)]
+pub struct SetReplicaEntityId {
+    pub uuid: ReplicaId,
+    pub entity_id: VolumeId,
+    pub node_id: NodeId,
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/replica.rs]: We have getters, so no need to make these pub

```suggestion
    uuid: ReplicaId,
    entity_id: VolumeId,
    node_id: NodeId,
```

File: control-plane/stor-port/src/transport_api/mod.rs [hunk 9]
```diff
@@ -184,6 +184,12 @@ impl From<HostNqnParseError> for ReplyError {
     }
 }
 
+impl From<uuid::Error> for ReplyError {
+    fn from(error: uuid::Error) -> Self {
+        Self::invalid_argument(ResourceKind::Unknown, "uuid", format!("{error:?}"))
+    }
+}
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/transport_api/mod.rs]: ResourceKind seems wrong, not sure you can do this as such

File: control-plane/agents/src/bin/core/node/wrapper.rs (hunk 7 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/node/wrapper.rs]: This refreshes resourceMap of the Nodewrapper right, replica map specifically? We probably dont need this step at all.We will update replica in the registry anyway if rpc call is Ok(). let nodewrapper update happen at its interval?

Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/node/wrapper.rs]: It refreshes everything which is not needed, only need to update the replica which gets returned.

File: control-plane/stor-port/src/transport_api/mod.rs (hunk 9 above)
Comment (abhilashshetty04) [reply control-plane/stor-port/src/transport_api/mod.rs]: Ill match result and return error. No need of this conversion.

**Comment by abhilashshetty04** [comment]
bors try

File: control-plane/agents/src/bin/core/controller/io_engine/v1/replica.rs [hunk 10]
```diff
@@ -155,6 +164,30 @@ impl crate::controller::io_engine::ReplicaApi for super::RpcClient {
...
+                    resource: ResourceKind::Replica,
+                    request: "set_replica_entity_id".to_string(),
+                    source: tonic::Status::data_loss(error.message()),
+                })
+            }
+            Err(error) => Err(error).context(GrpcRequestError {
+                resource: ResourceKind::Replica,
+                request: "set_replica_entity_id",
+            }),
+        };
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/replica.rs]:
```suggestion
        let rpc_replica = self
            .replica()
            .set_replica_entity_id(request.to_rpc())
            .await
            .context(GrpcRequestError {
                resource: ResourceKind::Replica,
                request: "set_replica_entity_id",
            })?;
        let replica = rpc_replica_to_agent(&rpc_replica.into_inner(), &request.node)?;
        Ok(replica)
```

File: control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs [hunk 11]
```diff
@@ -90,6 +92,15 @@ impl TryIoEngineToAgent for v1::replica::Replica {
                 uuid: self.uuid.to_owned(),
                 kind: ResourceKind::Replica,
             })?,
+            entity_id: match &self.entity_id {
+                None => None,
+                Some(entity_id) => Some(VolumeId::try_from(entity_id.clone()).map_err(|_| {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs]: I wonder if we should just set it to None if it's not a uuid?

File: control-plane/agents/src/bin/core/nexus/operations_helper.rs [hunk 12]
```diff
@@ -8,8 +8,12 @@ use crate::controller::{
     },
 };
 use agents::errors::SvcError;
+use tracing::log::error;
 
-use crate::controller::resources::operations::{ResourceOwnerUpdate, ResourceSharing};
+use crate::controller::{
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/nexus/operations_helper.rs]: Can you move these up together to line 1 and keep an empty line between agents and tracing

File: control-plane/agents/src/bin/core/nexus/operations_helper.rs [hunk 13]
```diff
@@ -133,6 +137,20 @@ impl OperationGuardArc<NexusSpec> {
         replica_state: &Replica,
         nexus_node: &NodeId,
     ) -> Result<ChildUri, SvcError> {
+        if replica_state.entity_id.is_none() {
+            let replica = registry.specs().replica(&replica_state.uuid).await?;
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/nexus/operations_helper.rs]: let's move the let mut replica from below up here so we don't have to do it twice.

File: control-plane/agents/src/bin/core/nexus/operations_helper.rs [hunk 14]
```diff
@@ -133,6 +137,20 @@ impl OperationGuardArc<NexusSpec> {
...
+            if let Some(vol) = replica.as_ref().owners.volume() {
+                let node = registry.node_wrapper(&replica_state.node).await?;
+                let set_entity_id = SetReplicaEntityId::new(
+                    replica_state.uuid.clone(),
+                    vol.clone(),
+                    replica_state.node.clone(),
+                );
+                if let Err(e) = node.set_replica_entity_id(&set_entity_id).await {
+                    error!("Failed to set entity_id for the replica {}", e.to_string());
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/nexus/operations_helper.rs]:
```suggestion
                    error!(replica.uuid = %replica_state.uuid, %error, "Failed to set the entity_id);
```

File: control-plane/agents/src/bin/core/controller/io_engine/v1/replica.rs [hunk 15]
```diff
@@ -88,16 +89,24 @@ impl crate::controller::io_engine::ReplicaListApi for super::RpcClient {
...
+                })
+            }
+            Err(error) => Err(error).context(GrpcRequestError {
                 resource: ResourceKind::Replica,
                 request: "create_replica",
-            })?;
-        let replica = rpc_replica_to_agent(&rpc_replica.into_inner(), &request.node)?;
-        Ok(replica)
+            }),
+        };
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/controller/io_engine/v1/replica.rs]: I think we can keep this as is, node wrapper is handling DataLoss anyway

File: control-plane/agents/src/bin/core/node/wrapper.rs [hunk 16]
```diff
@@ -1393,6 +1401,30 @@ impl ReplicaApi for Arc<tokio::sync::RwLock<NodeWrapper>> {
...
+            Err(SvcError::GrpcRequestError { source, .. })
+                if source.code() == tonic::Code::PermissionDenied =>
+            {
+                Err(SvcError::ReplicaSetXattrFailed {
+                    attribute: "entity_id".to_string(),
+                    replica: request.uuid().to_string(),
+                })
+            }
+            Err(error) => Err(error),
+        };
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/wrapper.rs]:
```suggestion
        match dataplane.set_replica_entity_id(request).await {
            Ok(replica) => {
                self.update_replica_state(Either::Insert(replica.clone()))
                    .await;
                Ok(replica)
            }
            Err(SvcError::GrpcRequestError { source, .. })
                if source.code() == tonic::Code::DataLoss =>
            {
                Err(SvcError::ReplicaSetXattrFailed {
                    attribute: "entity_id".to_string(),
                    replica: request.uuid().to_string(),
                })
            }
            Err(error) => Err(error),
        }
```

File: control-plane/agents/src/common/errors.rs [hunk 17]
```diff
@@ -394,6 +394,8 @@ pub enum SvcError {
     },
     #[snafu(display("Invalid property name '{property_name}' for the volume '{id}'"))]
     InvalidSetProperty { property_name: String, id: String },
+    #[snafu(display("Cannot set '{attribute}' on Replica '{replica}''s Lvol xattr"))]
+    ReplicaSetXattrFailed { attribute: String, replica: String },
```
Comment (tiagolobocastro) [inline control-plane/agents/src/common/errors.rs]: Let's not use XAttr, that's dataplane speak, control-plane does need to know, also we might have other backends soon..

```suggestion
    #[snafu(display("Cannot set entity_id '{entity_id}' on Replica '{replica}'"))]
    ReplicaSetEntityFailed { entity: String, replica: String },
```

File: control-plane/grpc/src/operations/replica/traits.rs [hunk 18]
```diff
@@ -375,6 +396,16 @@ impl CreateReplicaInfo for ValidatedCreateReplicaRequest {
         self.uuid.clone()
     }
 
+    fn entity_id(&self) -> Option<VolumeId> {
+        match self.inner.entity_id.clone() {
```
Comment (tiagolobocastro) [inline control-plane/grpc/src/operations/replica/traits.rs]: Isn't entity already volumeId?

File: control-plane/stor-port/src/types/v0/transport/replica.rs [hunk 19]
```diff
@@ -556,6 +559,40 @@ impl From<ReplicaOwners> for models::ReplicaSpecOwners {
...
+            node_id,
+        }
+    }
+
+    /// Returns uuid of the replica.
+    pub fn uuid(&self) -> &ReplicaId {
+        &self.uuid
+    }
+
+    /// Returns entity id of the replica, If set.
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/replica.rs]:
```suggestion
    /// Returns entity id of the replica.
```

File: control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs (hunk 11 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/io_engine/v1/translation.rs]: Yes , we dont validate UUID in dataplane :).

File: control-plane/agents/src/common/errors.rs (hunk 17 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/common/errors.rs]: How about ReplicaSetPropertyFailed { attribute: String, replica: String } ? We could reuse this if some other property set fails? attribute will have entity_id.

File: control-plane/agents/src/bin/core/controller/io_engine/v1/replica.rs (hunk 10 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/controller/io_engine/v1/replica.rs]: If we use context , Will it map the tonic status code to GrpcRequestError.source? We would need status code in wrapper to map it to SvcError:SetReplicaEntityId error

Comment (tiagolobocastro) [reply control-plane/agents/src/bin/core/controller/io_engine/v1/replica.rs]: No we don't as the status is already in `GrpcRequestError`

File: control-plane/agents/src/common/errors.rs (hunk 17 above)
Comment (tiagolobocastro) [reply control-plane/agents/src/common/errors.rs]: Sure, that seems good too

File: Cargo.lock
Comment (tiagolobocastro) [inline Cargo.lock:1]: this is already in tree, probably need to rebase

File: control-plane/agents/src/bin/core/node/wrapper.rs [hunk 20]
```diff
@@ -1283,6 +1283,14 @@ impl ReplicaApi for Arc<tokio::sync::RwLock<NodeWrapper>> {
 
                 Ok(replica)
             }
+            Err(SvcError::GrpcRequestError { source, .. })
+                if source.code() == tonic::Code::DataLoss =>
+            {
+                Err(SvcError::ReplicaSetPropertyFailed {
```
Comment (tiagolobocastro) [inline control-plane/agents/src/bin/core/node/wrapper.rs:1289]: Why do we need this additional mapping? Can't we just keep the grpc error as is?

File: control-plane/stor-port/src/types/v0/transport/replica.rs [hunk 21]
```diff
@@ -418,6 +420,7 @@ pub struct CreateReplica {
     pub name: Option<ReplicaName>,
     /// UUID of the replica.
     pub uuid: ReplicaId,
+    pub entity_id: Option<VolumeId>,
```
Comment (tiagolobocastro) [inline control-plane/stor-port/src/types/v0/transport/replica.rs:424]: missing doc comment

File: utils/dependencies
Comment (tiagolobocastro) [inline utils/dependencies:1]: Why is this showing as diff? Ah maybe because again rebase needed

**Review by tiagolobocastro — APPROVED** [review]

File: control-plane/agents/src/bin/core/node/wrapper.rs (hunk 20 above)
Comment (abhilashshetty04) [reply control-plane/agents/src/bin/core/node/wrapper.rs:1289]: It gives more context on why grpc failed with `Dataloss` status code. If it fails in `create_replica` we just mark replica for deletion.

**Review by Abhinandan-Purkait — APPROVED** [review]

**Comment by abhilashshetty04** [comment]
bors merge

**Comment by abhilashshetty04** [comment]
bors cancel

**Comment by abhilashshetty04** [comment]
bors merge

**Pull request merged**